napari/_vispy/layers/labels.py

```diff
@@ -120,12 +190,18 @@ def __init__(
         self,
         use_selection=False,
         selection=0.0,
+        collision=True,
     ):
         colors = ['w', 'w']  # dummy values, since we use our own machinery
         super().__init__(colors, controls=None, interpolation='zero')
-        self.glsl_map = direct_lookup_shader.replace(
-            '$use_selection', str(use_selection).lower()
-        ).replace('$selection', str(selection))
+        if collision:
+            self.glsl_map = direct_lookup_shader.replace(
+                '$use_selection', str(use_selection).lower()
+            ).replace('$selection', str(selection))
+        else:
+            self.glsl_map = direct_lookup_shade_without_collision.replace(
+                '$use_selection', str(use_selection).lower()
+            ).replace('$selection', str(selection))
```

Comment:
Two things I'm thinking of here:
1. these two shaders are basically identical except for the extra while loop for collision checking. Do we really need to maintain 2 separate shaders? I think it would be better to at least have a single shader with a bool uniform like `check_collision` which simply decides what to do.
2. I know this was already here before this PR, but we should probably use vispy's replacement logic instead of this hacky replace code. You should be able to do it with something like this:
```py
from vispy.visuals.shaders import Function

func = Function(direct_lookup_shader)
func['use_selection'] = use_selection
func['selection'] = selection
func['check_collision'] = collision
self.glsl_map = func
```

Reply:
I do not know. I do not exactly know how shader compilation works, how it cut dead code. In general the code executed on GPU do not like `if` clauses. 

The code replacement is to allow compiler to optimize code during compilation time and to not execute redundant checks. So if it is possible, then replacement should be used instead of providing variable value.

Reply:
As far as I know, this is only a problem if branching differs between fragments. If all fragments have the same branching (e.g: because of a bool uniform) then it should be compiled away.

As for the variables already used (`use_selection` and `selection`), vispy already replaces them inplace, except it has some extra checks to ensure everything is consistent, so the replacement there should be exactly identical in terms of final compilation result.

Reply:
I have checked. Even if vispy documentation claims that you are right this vispy feature is not working :P

Reply:
In what way it doesn't work? I have used it in other places before, so it should work in some way...

Reply:
(but this is not blocking for this PR, if you prefer to move forward)

Reply:
It declares a global variable but not do replace `$variable_name`. The shade compilation fails.

Reply:
This works as intended with just the `Function` object:

```py
from vispy.visuals.shaders import Function
low_disc_lookup_shader = """
uniform sampler2D texture2D_LUT;

vec4 sample_label_color(float t) {
    float phi_mod = 0.6180339887498948482;  // phi - 1
    float value = 0.0;
    float margin = 1.0 / 256;

    if (t == 0) {
        return vec4(0);
    }

    if (($use_selection) && ($selection != t)) {
        return vec4(0);
    }

    value = mod((t * phi_mod + $seed), 1.0) * (1 - 2*margin) + margin;

    return texture2D(
        texture2D_LUT,
        vec2(0.0, clamp(value, 0.0, 1.0))
    );
}
"""
func = Function(low_disc_lookup_shader)
func['seed'] = 1
func['selection'] = 2
func['use_selection'] = True
print(func.compile())
```

Something must be broken in how the `glsl_map` is being used...